### ISSUE TITLE
Elsa 2 - If/Else weird

### ISSUE THREAD

**Issue by riatdyl** [elsa-workflows/elsa-core#941]
Hello,

The If/Else has three branches : True, False and Done

If I have a True outcome, it executes immediately and then the Done branch is executed.

But if I have a False outcome, it's waiting on the Done branch before execution.

**Comment by sfmskywalker** [comment]
I presume this is Elsa 1. It does sound like an issue, since the intent is for the Done branch to execute lastly.

With Elsa 2, it does work correctly, albeit slightly different compared to Elsa 1. In Elsa 2, the Done branch is executed after the entire True or False branch has finished executing. This is in contrast with Elsa 1, where the Done branch executes immediately after (or apparently before) the first activity connected to True/False completes.

Given that the very reason of Elsa 2's existence is to fix the issues in Elsa 1, I am going to mark this issue as "won't fix", since it does work correctly in Elsa 2.

Elsa 2 will be released next week 🤞🏻

**Comment by riatdyl** [comment]
I'm actually using Elsa 2.

**Comment by sfmskywalker** [comment]
Then that's indeed a strange issue. I tried to reproduce, but it works OK for me. Could you create a repro that I can look at?

**Comment by riatdyl** [comment]
Okay I managed to reproduce the problem.

![image](https://user-images.githubusercontent.com/43779018/117446335-39598a00-af3c-11eb-8375-d07cb433b5d8.png)
Here, the workflow is waiting on the second "ReadLine" (the one in the "Done" branch) before executing the "HTTP Response" in the True or False branch.

It means that the "Done" branch is executing before the others branches are executed.

And it's weird because in the "HTTP Response", I display the content of the Variable. The value of the variable will be the value of the first "ReadLine" (the one before the "If/Else" activity).

**Comment by sfmskywalker** [comment]
How can you tell that the second `ReadLine` executes before the two `HttpResponse` activities connected to the True and False branches?

**Comment by riatdyl** [comment]
Because in my browser, the HTTP Response is displaying after I entered a second input in the console.

**Comment by riatdyl** [comment]
Hmm, it's maybe just a diplay problem.

I looked in the Workflow instances and it looks like this after the workflow is executed completely : 
![image](https://user-images.githubusercontent.com/43779018/117447908-5db66600-af3e-11eb-8049-93120e9fa780.png)

**Comment by sfmskywalker** [comment]
I think I understand.

What happens is that after the _True_ or _False_ branch execute, the `If/Else` activity is scheduled again so that it gets a chance to schedule the *Done* outcome. This is why in the activity log this activity appears again, since technically, it executed again. This process is called *unwinding*.

I realize this is confusing if you don't know about this internal implementation detail and I am thinking about improvements in this area.

**Comment by riatdyl** [comment]
>But the page is loading the content of the HTTP Response after the second Readline
When I was talking about the second Readline, I was not talking about the log.

The fact is that the HTTP Response shows me the response on the browser only after I write something in the second Readline. Otherwise it will just wait. So that's why I think the Done is executing before the True and False.

PS : I'm sorry, I speak french and I do my best to be understandable ^^

**Comment by sfmskywalker** [comment]
Right. You see, it makes perfect sense that you only see the HTTP response after you write something in the second ReadLine. What happens is that first a response is written to the response stream. But before that response gets flushed to the client, the ReadLine activity needs to complete. Only after you write something will that activity complete, and will the response arrive in the browser.

I didn’t notice you were French; you’re English is quite good :)

**Comment by riatdyl** [comment]
Okay thank you. I'm sorry for my misunderstanding ^^'

I'm from Switzerland and it's nice to hear that I have correct English :)

**Comment by sfmskywalker** [comment]
You got it. No worries, we’re all learning 💪🏻
